assignments/burmanm/week-2.scm

```diff
@@ -0,0 +1,394 @@
+	  (* base (expmod base (- exp 1) m))
+	  m))))
+
+(define (fermat-test n)
+  (define (try-it a)
+    (= (expmod a n n) a))
+  (try-it (+ 1 (random (- n 1)))))
+(define (fast-prime? n times)
+  (cond ((= times 0) true)
+	((fermat-test n) (fast-prime? n (- times 1)))
+	(else false)))
+
+;; Use our modified version of timed-prime-test to reduce printing to only primes
+
+(define (timed-prime-test n)
+  (start-prime-test n (runtime)))
+(define (start-prime-test n start-time)
+  (if (fast-prime? n 100)
+      (report-prime (- (runtime) start-time) n)
+      #f))
+(define (report-prime elapsed-time n)
+  (newline)
+  (display n)
+  (display " *** ")
+  (display elapsed-time)
+  #t)
+
+;; Use (find-next-three (round->exact n)) to get it working with large numbers..
+
+;; The time to calculate fermat numbers does not seem to increase at all and is less than the precision of
+;; the runtime. Replacing runtime with real-time-clock (mit-scheme) gives us a better indication:
+
+(define (timed-prime-test n)
+  (start-prime-test n (real-time-clock)))
+(define (start-prime-test n start-time)
+  (if (fast-prime? n 100)
+      (report-prime (- (real-time-clock) start-time) n)
+      #f))
+
+
+;; 10 error> (find-next-three (round->exact 1e40))
+
+;; 10000000000000000303786028427003666890753 *** 53
+;; 10000000000000000303786028427003666891041 *** 30
+;; 10000000000000000303786028427003666891101 *** 32
+;; ;Value: 3
+
+;; 10 error> (find-next-three (round->exact 1e300))
+
+;; 1000000000000000052504760255204420248704468581108159154915854115511802457988908195786371375080447864043704443832883878176942523235360430575644792184786706982848387200926575803737830233794788090059368953234970799945081119038967640880074652742780142494579258788820056842838115669472196386865459400540381 *** 502
+;; 1000000000000000052504760255204420248704468581108159154915854115511802457988908195786371375080447864043704443832883878176942523235360430575644792184786706982848387200926575803737830233794788090059368953234970799945081119038967640880074652742780142494579258788820056842838115669472196386865459400542127 *** 510
+;; 1000000000000000052504760255204420248704468581108159154915854115511802457988908195786371375080447864043704443832883878176942523235360430575644792184786706982848387200926575803737830233794788090059368953234970799945081119038967640880074652742780142494579258788820056842838115669472196386865459400543303 *** 496
+;; ;Value: 3
+
+;; 10 error> (find-next-three (round->exact 1e150))
+
+;; 999999999999999980835596172437374590573120014030318793091164810154100112203678582976298268616221151962702060266176005440567032331208403948233373515999 *** 151
+;; 999999999999999980835596172437374590573120014030318793091164810154100112203678582976298268616221151962702060266176005440567032331208403948233373516469 *** 151
+;; 999999999999999980835596172437374590573120014030318793091164810154100112203678582976298268616221151962702060266176005440567032331208403948233373517033 *** 150
+;; ;Value: 3
+
```

Comment:
at 1e300 you measure ~500, and at 1e150 you measure 150. If it was logarithmic growth, shouldn't 1eX = O(t) be 1e2X = O(2t) ?

Reply:
Yes, but then that would be 300 ~ 500. Even if the growth would actually be closer to O(2 log n), it is still simplified to form O(log n). It's only the exponential behavior that's taken into account in big O notation.

Reply:
thanks for the explanation.